contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,334 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {IERC20} from "@openzeppelin/contracts-4.8.0/token/ERC20/IERC20.sol";
+import {IERC20Metadata} from "@openzeppelin/contracts-4.8.0/token/ERC20/extensions/IERC20Metadata.sol";
+import {SafeERC20} from "@openzeppelin/contracts-4.8.0/token/ERC20/utils/SafeERC20.sol";
+import {Math} from "@openzeppelin/contracts-4.8.0/utils/math/Math.sol";
+
+import {IPrivatePool} from "./interfaces/IPrivatePool.sol";
+
+/// @title Private pool for concentrated liquidity
+/// @notice Allows LP to offer fixed price quote in private pool to bridgers for tighter prices
+/// @dev Obeys constant sum P * x + y = D curve, where P is fixed price and D is liquidity
+/// @dev Functions use same signatures as Swap.sol for easier integration
+contract PrivatePool is IPrivatePool {
+    using SafeERC20 for IERC20;
+
+    uint256 internal constant wad = 1e18;
+    uint256 internal constant PRICE_BOUND = 0.001e18; // 10 bps in wad
+
+    uint256 public constant PRICE_MIN = wad - PRICE_BOUND; // 1 - 10bps in wad
+    uint256 public constant PRICE_MAX = wad + PRICE_BOUND; // 1 + 10bps in wad
+    uint256 public constant FEE_MAX = 0.001e18; // 10 bps in wad
+
+    address public immutable factory;
+    address public immutable owner;
+
+    address public immutable token0; // base token
+    address public immutable token1; // quote token
+
+    uint256 internal immutable token0Decimals;
+    uint256 internal immutable token1Decimals;
+
+    uint256 public P; // fixed price param: amount of token1 per amount of token0 in wad
+    uint256 public fee; // fee charged on swap; acts as LP's bid/ask spread
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "!owner");
+        _;
+    }
+
+    modifier onlyToken(uint8 index) {
+        require(index <= 1, "invalid token index");
+        _;
+    }
+
+    modifier deadlineCheck(uint256 deadline) {
+        require(block.timestamp <= deadline, "block.timestamp > deadline");
+        _;
+    }
+
+    modifier hasQuote() {
+        require(P > 0, "invalid quote");
+        _;
+    }
+
+    event Quote(uint256 price);
+    event NewSwapFee(uint256 newSwapFee);
+    event TokenSwap(address indexed buyer, uint256 tokensSold, uint256 tokensBought, uint128 soldId, uint128 boughtId);
+    event AddLiquidity(
+        address indexed provider,
+        uint256[] tokenAmounts,
+        uint256[] fees,
+        uint256 invariant,
+        uint256 lpTokenSupply
+    );
+    event RemoveLiquidity(
+        address indexed provider,
+        uint256[] tokenAmounts,
+        uint256[] fees,
+        uint256 invariant,
```

Comment:
I think it makes sense to replicate the `Swap.sol` event structure here, so that the off-chain event parser won't require custom logic for these pools.

```solidity
    event RemoveLiquidity(
        address indexed provider,
        uint256[] tokenAmounts,
        uint256 lpTokenSupply
    );
```

Reply:
Sounds good! Will fix